Code_Exercises/CMakeLists.txt

```diff
@@ -8,32 +8,102 @@
   see <http://creativecommons.org/licenses/by-sa/4.0/>.
 ]]
 
-function(add_sycl_executable prefix source)
-  add_executable("${prefix}_${source}" "${source}.cpp")
-  target_compile_definitions("${prefix}_${source}" PUBLIC
-    -DSYCL_LANGUAGE_VERSION=2020)
-  if(${SYCL_ACADEMY_USE_COMPUTECPP})
-    target_compile_definitions("${prefix}_${source}" PUBLIC -DSYCL_ACADEMY_USE_COMPUTECPP)
+message(STATUS ">> CMake Exercises")
+message(STATUS "SYCL_ACADEMY_BUILD_EXERCISES ${SYCL_ACADEMY_BUILD_EXERCISES}")
+
+# CMake function to build a ComputeCPP configuration and create executable binaries
+# from ComputeCPP libraries. 
+function( add_sycl_executable_computecpp prefix source )
+  if(NOT DEFINED ${CMAKE_BUILD_TYPE})
+    set(CMAKE_BUILD_TYPE "DEBUG")
   endif()
+
+  add_executable("${prefix}_${source}" "${source}.cpp")
+  target_compile_definitions("${prefix}_${source}" PUBLIC -DSYCL_LANGUAGE_VERSION=2020)
+  target_compile_definitions("${prefix}_${source}" PUBLIC -DSYCL_ACADEMY_USE_COMPUTECPP)
   target_include_directories("${prefix}_${source}" PRIVATE
-    ${PROJECT_SOURCE_DIR}/Utilities/include ${PROJECT_SOURCE_DIR}/External/stb)
-  target_link_libraries("${prefix}_${source}" PRIVATE Threads::Threads)
+    ${PROJECT_SOURCE_DIR}/Utilities/include 
+    ${PROJECT_SOURCE_DIR}/External/stb)
+  target_link_libraries("${prefix}_${source}" PRIVATE Threads::Threads) 
   target_link_libraries("${prefix}_${source}" PUBLIC Catch2::Catch2)
-  set_target_properties("${prefix}_${source}" PROPERTIES CXX_STANDARD 17
-    CXX_STANDARD_REQUIRED ON)
-
-  add_sycl_to_target(
-    TARGET "${prefix}_${source}"
-    SOURCES "${source}.cpp"
-  )
-    
+  set_target_properties("${prefix}_${source}" PROPERTIES CXX_STANDARD 17 CXX_STANDARD_REQUIRED ON)  
   add_test(${prefix}_${source} ${prefix}_${source})
 endfunction()
 
-function(add_sycl_executable_dpcpp prefix source)
-  # Not implemented
+# CMake function to build a hipSYCL configuration and create executable binaries
+# from hipSYCL libraries. 
+function( add_sycl_executable_hipSYCL prefix source )
+  if(NOT DEFINED ${CMAKE_BUILD_TYPE})
+	  set(CMAKE_BUILD_TYPE "DEBUG")
+  endif()
+
+  add_executable("${prefix}_${source}" "${source}.cpp")
+  target_compile_definitions("${prefix}_${source}" PUBLIC -DSYCL_LANGUAGE_VERSION=2020)
+  target_include_directories("${prefix}_${source}" PRIVATE
+    ${PROJECT_SOURCE_DIR}/Utilities/include 
+    ${PROJECT_SOURCE_DIR}/External/stb)
+  target_link_libraries("${prefix}_${source}" PRIVATE Threads::Threads) 
+  target_link_libraries("${prefix}_${source}" PUBLIC Catch2::Catch2)
+  set_target_properties("${prefix}_${source}" PROPERTIES CXX_STANDARD 17 CXX_STANDARD_REQUIRED ON)  
+endfunction()
+
+# CMake function to build a DPC++ configuration and create executable binaries
+# from oneAPI libraries. 
+function( add_sycl_executable_dpcpp prefix source )
```

Comment:
So looking at the stuff that we have here, maybe we can unify the dpcpp and (the artist formerly known as) hipsycl code, since they behave the same way (i.e. `clang++ -fsycl $src_file.cpp`). That would cut down on our code a lot, especially with the ComputeCpp stuff being removed. (Originally `add_sycl_to_target` was a ComputeCppism that Intel have made some effort to copy, but it's not needed in the single-compiler model really. It made sense when we were adding a separate compiler pass then messing with the way that the user sources were compiled.)

Reply:
These changes could be applied in the wip_ComputeCPP_refsToAllRemove branch. Changes from this branch have been applied in the latter branch wip_ComputeCPP_refsToAllRemove. Split out the functions into 3 distinct functions, one for each compiler type so be be very clear as to what is being done to which, no if then else's  noise. Make the functions truly independent of each other. Makes maintainability easier. Reduce risk of inadvertent changes to one build effecting another build (which may not be tested for at the time).

Reply:
I understand that but as much as is possible I'd like the same treatment for all of the different implementations. It seems like AdaptiveCpp also supports the `add_sycl_to_target` functionality so it might be possible to unify all the build functions into one.

Reply:
AdaptiveCpp build cannot be tested so leaving as is. Generate a JIRA task to make and test hipSYCL / AdaptiveCpp build.